src/sdk/namespace_fs.js

```diff
@@ -1759,6 +1797,197 @@ class NamespaceFS {
             throw new RpcError('BAD_REQUEST', 'Bad Request');
         }
     }
+
+    // delete latest version - 
+    // 1. if (key) safe move key .versions/key_{version_id}
+    // 2. else latest version is a delete marker / doesn't exist - nothing to move
+    // 3. create delete marker and move it to .versions/key_{version_id}_delete_marker
+    // delete version_id - 
+    // 1. unlink key
+    // 2. find max past version
+    // 3. if deleted version is latest && max_past_version is a regular version (not a delete marker) -
+    //    3.1. move it file path (to be latest)
+    // 4. if deleted version is a delete marker && max_past_version older than the deleted version 
+    //    4.1. move it file path (to be latest)
+    async _delete_object_versioning_enabled(fs_context, file_path, params) {
+        return params.version_id ?
+            this._delete_version_id(fs_context, file_path, params) :
+            this._delete_latest_version(fs_context, file_path, params);
+    }
+
+    // iterate over the key's versions array
+    // if version is undefined, delete latest
+    // if version exists - unlink version
+    // promote second latest to latest 
+    async _delete_objects_versioning_enabled(fs_context, key, versions) {
+        let res = [];
+        for (const version_id of versions) {
+            try {
+                if (version_id) {
+                    const file_path = await this._find_version_path(fs_context, { key, version_id });
+                    const version_info = await this._get_version_info(fs_context, file_path);
+                    if (!version_info) {
+                        res.push({});
+                        continue;
+                    }
+                    await nb_native().fs.unlink(fs_context, file_path);
+                    res.push({ deleted_delete_marker: version_info.delete_marker });
+                } else {
+                    const file_path = this._get_file_path({ key });
+                    const version_res = await this._delete_latest_version(fs_context, file_path, { key, version_id });
+                    res.push(version_res);
+                }
+            } catch (err) {
+                res.push({ err_code: err.code, err_message: err.message });
+            }
+        }
+        // delete .versions/ if it's empty
+        const file_path = this._get_version_path(key);
+        await this._delete_path_dirs(file_path, fs_context);
+        await this._promote_version_to_latest(fs_context, { key }, undefined);
```

Comment:
as we may delete the last version here we need to try and delete the path after the promotion